### PR TITLE
change the timing of sorting logits

### LINKED ISSUE #1179: Is Sampler forward() working properly now?
Hello.
I'm using the vLLM very well.
But in a recent commit, I find some code that may have bugs.

I suspect there is a bug in the way of applying sampling_params in the vllm/model_executor/layers/sampler.py.

Watching a _prune_hidden_states function , It seems that logits is sorted by the SamplingType.
However, when getting the sampling parameters, those ​​are just ordered by the input_metadata.seq_groups. (e.g. _get_penalties)
I can't find a way to match these two in a step of applying parameters. (e.g. _apply_penalties)
It may be happened that sampling parameters are applied to wrong logits.

Please let me know if I misunderstood anything.

### PR DESCRIPTION
**Pull request by yhlskt23**
This PR fixes #1179 .

There is a bug in the way sampling parameters are applied to logits.
The logits are sorted before the parameters are applied, so there are in danger of using other input's parameters.

For example, assume that the following 4 inputs are in a batch, in this order : 
<greedy, frequency penalty = 0.0>, <random, frequency penalty = 0.0>, <greedy, frequency penalty = 1.0>, <random, frequency penalty = 1.0>

Before applies the frequency penalty, logits are sorted by sampling type, such as GREEDY, RANDOM, or BEAM
Logits : <**greedy**, frequency penalty = 0.0>, <**greedy**, frequency penalty = 1.0>, <**random**, frequency penalty = 0.0>, <**random**, frequency penalty = 1.0>

However, the sampling parameters do not change their order.
Parameters : <greedy, **frequency penalty = 0.0**>, <random, **frequency penalty = 0.0**>, <greedy, **frequency penalty = 1.0**>, <random, **frequency penalty = 1.0**>

Therefore, the frequency penalty is used in the wrong order, which changes the final result to something unexpected.
results : <**greedy**, **frequency penalty = 0.0**>, <**greedy**, **frequency penalty = 0.0**>, <**random**, **frequency penalty = 1.0**>, <**random**, **frequency penalty = 1.0**>

I change the timing of the sorting to just the beginning of the final sampling function(=_sample)

### CODE REVIEW

**Comment by zhuohan123** [comment]
Thanks for reporting this. This is indeed a bug. I'll test and review the PR later today.

**Review by zhuohan123 — APPROVED** [review]
LGTM! Thanks for the fix. I removed the unused `category_start_idx` as well. The performance of the following benchmark:
```
python benchmark_throughput.py --backend vllm --model huggyllama/llama-7b --dataset ../../data/ShareGPT_V3_unfiltered_cleaned_split.json --
num-prompts 2000
```
dropped from 8.70 reqs/s to 8.64 reqs/s on my A100 80G machine. Likely due to the newly added indexing operator. Will probably fix this later.
